File: ws/src/main/scala/de/thm/ii/fbs/controller/LegalController.scala
package de.thm.ii.fbs.controller

import de.thm.ii.fbs.controller.exception.{ForbiddenException, ResourceNotFoundException}
import de.thm.ii.fbs.services.persistance.UserService
import de.thm.ii.fbs.services.security.AuthService
import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation._

import scala.io.Source
import scala.util.{Success, Using}

/**
  * Manages legally required information.
  */
@RestController
@CrossOrigin
@RequestMapping(path = Array("/api/v1/legal"))
class LegalController {
  @Autowired
  private implicit val userService: UserService = null
  @Autowired
  private val authService: AuthService = null

  /**
    * Load the markdown files from file system and display it
    * @param filename markdown file name
    * @param request contain request information
    * @return markdown file content
    */
  @GetMapping(value = Array("/{filename}"))
  @ResponseBody
  def legalTexts(@PathVariable filename: String, request: HttpServletRequest): Map[String, String] = {
    val filePath = filename match {
      case "impressum" => Some("impressum.md")
      case "privacy-text" => Some("privacy_text.md")
      case _ => None
    }
    filePath match {
      case Some(path) =>
        Using(Source.fromResource(path)) {s => s.mkString}.map(text => Map("markdown" -> text)) match {
          case Success(res) => res
          case _ => throw new Exception(s"Could not read $path")
        }
      case None => throw new ResourceNotFoundException
    }
  }

  /**
    * Get the information if a user has accepted the terms of use
    * @param uid User id
    * @param req request
    * @param res response
    * @return response content
    */
  @GetMapping(value = Array("/termsofuse/{uid}"))
  @ResponseBody
  def getTermsOfUseAcceptanceStatus(@PathVariable uid: Int, req: HttpServletRequest, res: HttpServletResponse): Map[String, Boolean] = {
    val user = authService.authorize(req, res)
    user.id match {
      case `uid` => Map("accepted" -> userService.getPrivacyStatusOf(user.id))
      case _ => throw new ForbiddenException()
    }
  }

  /**
    * Accept the terms of use
    * @param uid User id
    * @param req request
    * @param res response
    */
  @PutMapping(value = Array("/termsofuse/{uid}"))
  def acceptTermsOfUse(@PathVariable uid: Int, req: HttpServletRequest, res: HttpServletResponse): Unit = {
    val user = authService.authorize(req, res)
    user.id match {
      case `uid` => userService.updateAgreementToPrivacyFor(user.id, agreed = true)
      case _ => throw new ForbiddenException()
    }
  }
}

File: modules/fbs-identity-service/api/src/main/kotlin/de/thm/ii/fbs/fbs_identity_service/controller/LegalController.kt
package de.thm.ii.fbs.fbs_identity_service.controller

import de.thm.ii.fbs.fbs_identity_service.model.user.User
import de.thm.ii.fbs.fbs_identity_service.service.auth.CurrentUserService
import de.thm.ii.fbs.fbs_identity_service.service.user.UserService
import org.springframework.core.io.ClassPathResource
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping("/api/v1/legal")
class LegalController(private val userService: UserService, private val currentUserService: CurrentUserService) {

    @GetMapping("/{filename}")
    fun legalTexts(@PathVariable filename: String): Map<String, String> {
        val resourceName = when (filename) {
            "impressum" -> "impressum.md"
            "privacy-text" -> "privacy_text.md"
            else -> throw ResponseStatusException(HttpStatus.NOT_FOUND)
        }

        val text = ClassPathResource(resourceName)
            .inputStream
            .bufferedReader()
            .use { it.readText() }

        return mapOf("markdown" to text)
    }

    @GetMapping("/termsofuse/{uid}")
    fun getTermsOfUseAcceptanceStatus(@PathVariable uid: Long): Map<String, Boolean> {
        val user = requireCurrentUser(uid)

        return mapOf("accepted" to userService.getPrivacyStatusOf(user.id))
    }

    @PutMapping("/termsofuse/{uid}")
    fun acceptTermsOfUse(@PathVariable uid: Long) {
        val user = requireCurrentUser(uid)
        userService.updateAgreementToPrivacyFor(user.id, true)
    }

    private fun requireCurrentUser(uid: Long): User {
        val user = currentUserService.getCurrentUser()
            ?: throw ResponseStatusException(HttpStatus.UNAUTHORIZED)

        if (user.id != uid) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN)
        }

        return user
    }
}